### PR TITLE
add milopy

### PR DESCRIPTION
**Pull request by xinyuejohn**
**PR Checklist**

-   [ ] Referenced issue is linked
-   [ ] If you've fixed a bug or added code that should be tested, add tests!
-   [ ] Documentation in `docs` is updated

**Description of changes**

**Technical details**

**Additional context**

### CODE REVIEW

File: .flake8 [hunk 1]
```diff
@@ -2,7 +2,7 @@
 select = B,B9,C,E,F,N,W
 ignore = E203,E501,RST201,RST203,RST301,W503,D100,F401,B950,N806,N803,F402
 max-line-length = 120
-max-complexity = 15
+max-complexity = 20
```
Comment (Zethson) [inline .flake8:5]: Can we tone this down to 15 again?

**Review by Zethson — CHANGES_REQUESTED** [review]
Nice!

1. I removed the manual type hints that you added to the docstrings. These are automatically inferred by Sphinx extensions. I also renamed milopy_mdata to just mdata.
2. It would be great if every method worked with a MuData object. I noticed that atm it is slightly inconsistent with some methods working with AnnData objects only. An example would be the `nhood` method.
3. I made the rpy2 dependency optional because it can only be installed if R is available.
4. The documentation build is failing. I really don't get why, but it would be great if you could look into it. The builds before this PR seem to succeed.

**Comment by Zethson** [comment]
@xinyuejohn well done! Could you please ensure that https://github.com/theislab/pertpy-tutorials/pull/4 is also in a perfect state? Then I can see how it works and if required complain about a few things.

**Comment by xinyuejohn** [comment]
@Zethson I updated the notebook and the functions accordingly. I think it is now well prepared.

File: pertpy/plot/_milopy.py [hunk 2]
```diff
@@ -0,0 +1,228 @@
...
+            hue="is_signif",
+            palette=["grey", "black"],
+            orient="h",
+            alpha=0.5,
+        )
+        plt.legend(loc="upper left", title=f"< {int(alpha*100)}% SpatialFDR", bbox_to_anchor=(1, 1), frameon=False)
+        plt.axvline(x=0, ymin=0, ymax=1, color="black", linestyle="--")
+
+    @staticmethod
+    def nhood_counts_by_cond(mdata: MuData, test_var: str, subset_nhoods: list = None, log_counts: bool = False):
```
Comment (emdann) [inline pertpy/plot/_milopy.py:194]: This method doesn't work: it's still geared to have nhoods in obs and samples in var, as in old milopy implementation.

**Review by emdann — CHANGES_REQUESTED** [review]
Thank you @xinyuejohn for the great work here! I tried to go through all the main functions, but if there's some bit that you'd like me to take a look at in particular let me know, in case I've missed it.

**Comment by emdann** [comment]
Also: what's the plan for installation of the R dependencies here? 😁 Will it be two lines in the installation docs or can these be added as extras?

**Comment by Zethson** [comment]
> Also: what's the plan for installation of the R dependencies here? 😁 Will it be two lines in the installation docs or can these be added as extras?

We can add the R dependencies as optional dependencies. However, we cannot actually add edgeR as a dependency to the package. That's also why we're hoping to have the batchGLM support asap....

**Comment by xinyuejohn** [comment]
> Thank you @xinyuejohn for the great work here! I tried to go through all the main functions, but if there's some bit that you'd like me to take a look at in particular let me know, in case I've missed it.

Thank you Emma for helping review my changes! It really helped a lot.

File: pertpy/plot/_milopy.py (hunk 2 above)
Comment (xinyuejohn) [reply pertpy/plot/_milopy.py:194]: Hi Emma, could you please explain what's the idea behind this function? At the moment I don't have clue of how to revise it to fit it into the current setting. Especially this part: 
```
if subset_nhoods is None:
    subset_nhoods = nhood_adata.obs_names

pl_df = pd.DataFrame(nhood_adata[subset_nhoods].X.A, columns=nhood_adata.var_names).melt(
    var_name=nhood_adata.uns["sample_col"], value_name="n_cells"
)
pl_df = pd.merge(pl_df, nhood_adata.var)
```

Thanks a lot!

Comment (emdann) [reply pertpy/plot/_milopy.py:194]: This function is to visualize the cell abundance for each sample, grouped by the covariate of interest. So it's a way to sanity check that the predicted enrichment/depletions make sense and what sort of effect size is captured.

Looking back at this the problem doesn't seem to be w the function, but it was complaining about the gastrulation dataset having integers in `mdata['milo_compositional'].obs` columns, so maybe it just needs to be a bit more robust (and to change the parameter names, cos `test_var` is actually a column in `mdata['milo_compositional'].obs`). To give you an ides of the output, on the tutorial dataset:
```
## Get ixs of enriched nhoods 
pl_nhoods = mdata['milo_compositional'].var_names[
    (mdata['milo_compositional'].var['SpatialFDR'] < 0.1) &
    (mdata['milo_compositional'].var['logFC'] > 0)
]

mdata['milo_compositional'].obs['sample'] = mdata['milo_compositional'].obs['sample'].astype('str')
mdata['milo_compositional'].obs['stage_continuous'] = mdata['milo_compositional'].obs['stage_continuous'].astype('category')
pt.pl.milo.nhood_counts_by_cond(mdata, test_var = 'stage_continuous', subset_nhoods = pl_nhoods)
``` 
Screenshot 2022-11-03 at 12 39 54

So the _subset_nhoods_ parameter is used to select the nhoods that you want to plot

> ```
> pl_df = pd.DataFrame(nhood_adata[subset_nhoods].X.A, columns=nhood_adata.var_names).melt(
>     var_name=nhood_adata.uns["sample_col"], value_name="n_cells"
> )
> pl_df = pd.merge(pl_df, nhood_adata.var)
> ```
> 
This bit of code makes a "long" dataframe from the cell abundance matrix (`nhood_adata.X`), then joins this dataframe with the covariate info for each sample stored in `nhood_adata.var`.

Does this clarify?

Comment (xinyuejohn) [reply pertpy/plot/_milopy.py:194]: Thanks for your clarification! It's now much clear to me.

File: pertpy/tools/_milopy.py [hunk 3]
```diff
@@ -0,0 +1,590 @@
...
+
+        nhoods = knn_graph[:, refined_vertices]
+        adata.obsm["nhoods"] = nhoods
+
+        # Add ixs to adata
+        adata.obs["nhood_ixs_random"] = adata.obs_names.isin(adata.obs_names[random_vertices])
+        adata.obs["nhood_ixs_refined"] = adata.obs_names.isin(adata.obs_names[refined_vertices])
+        adata.obs["nhood_ixs_refined"] = adata.obs["nhood_ixs_refined"].astype("int")
+        adata.obs["nhood_ixs_random"] = adata.obs["nhood_ixs_random"].astype("int")
+        # Store info on neighbor_key used
```
Comment (Zethson) [inline pertpy/tools/_milopy.py:151]:
```suggestion
```

**Review by Zethson — APPROVED** [review]
Thank you very much! A few minor things but I'll fix them at some point by myself.

We can also still adapt things if @emdann has more requests after she overhauled the tutorial.

**Pull request merged**